apache2/waf_logging/waf_log_util_internal.h

```diff
@@ -1,9 +1,100 @@
 #ifndef _WAF_LOG_UTIL_INTERNAL_HEADER
 #define _WAF_LOG_UTIL_INTERNAL_HEADER
 
-#include <google/protobuf/util/json_util.h>
-#include"waf_format.pb.h"
 #include <fstream>
+#include <unordered_map>
 #include <exception>
+#include "../../jsoncons/include/jsoncons/json.hpp"
+#include "../../jsoncons/include/jsoncons/json_options.hpp"
+
+// For convenience
+using jsoncons::ojson;
+using jsoncons::json_options;
+using jsoncons::line_split_kind;
+
+namespace ns {
+
+    struct detail
+    {
+        std::string message;
+        std::string data;
+        std::string file;
+        std::string line;
+    };
+
+    struct property
+    {
+        std::string instanceId;
+        std::string clientIp;
+        std::string clientPort;
+        std::string requestUri;
+        std::string ruleSetType;
+        std::string ruleSetVersion;
+        std::string ruleId;
+        std::string message;
+        std::string action;
+        std::string site;
+        detail details;
+        std::string hostname;
+        std::string transactionId;
+    };
+
+    class waf_log_object
+    {
+        std::string resourceId;
+        std::string operationName;
+        std::string category;
+        property properties;
+
+        // Make json_type_traits specializations friends to give accesses to private members
+        JSONCONS_TYPE_TRAITS_FRIEND;
+    public:
+        waf_log_object(){
+        }
+
+        waf_log_object(const std::string& resourceId, const std::string& operationName, const std::string& category, const property& properties)
```

Comment:
You should not need this constructor if you declare `waf_log_object` to be a `struct` instead of a `class` and initialize it using curly braces which will trigger aggregate initialization.

Reply:
will change to struct